### DISCUSSION TITLE
Random logout after refreshing quickly

### DISCUSSION THREAD

**Discussion by peterhijma** [pocketbase/pocketbase#3008] [Q&A]
I have a hard time debugging a problem where if I refresh very quickly a few times, the user gets logged out. I'm not sure if the problem lies within Nuxt (nuxt itself or my code), or pocketbase. The Nuxt plugin I use is the same as the example in the Pocketbase JS SDK readme:

```
export default defineNuxtPlugin(async () => {
  const config = useAppConfig();

  const pb = new PocketBase(config.pocketbase.url);

  const cookie = useCookie<CookieType>("pb_auth", {
    path: "/",
    secure: process.env.NODE_ENV === "production",
    sameSite: "strict",
    httpOnly: false,
    maxAge: 604800,
  });

  pb.authStore.save(cookie.value?.token, cookie.value?.model);

  pb.authStore.onChange(() => {
    cookie.value = {
      token: pb.authStore.token,
      model: pb.authStore.model,
    };
  });

  try {
    pb.authStore.isValid && (await pb.collection("users").authRefresh());
  } catch (e) {
    pb.authStore.clear();
  }
  return {
    provide: { pb },
  };
});
```

At some point, `pb.authStore.clear();` get's called. So `pb.authStore.isValid && (await pb.collection("users").authRefresh());` fails.

**Comment by peterhijma** [comment]
It does go to `pb.authStore.clear();`

**Reply by ganigeorgiev** [reply]
Could you share a more complete code sample showing where the plugin is invoked?

The Nuxt3 example could be just no longer valid as I've written it while v3 was still in beta.

I'll try to review the nuxt3 docs sometime later this week and will update the example if needed.

**Reply by peterhijma** [reply]
Just a page containing this:

```
<script setup>
const { $pb } = useNuxtApp();

const { data, refresh } = await useAsyncData("foo", async () => {
  const records = await $pb.collection("foo").getList(null, null);
  return structuredClone(records).items;
});
</script>

<template>
...
</template>
```

Also if I remove `pb.authStore.isValid && (await pb.collection("users").authRefresh());`, the problem does not occur.

It feels like some kind of race condition? 🤔

**Reply by peterhijma** [reply]
For your info, I'm refreshing by just holding SHIFT-COMMAND-R (MacOS, Brave browser). So.. it does fire a lot of refreshes very fast. (If I remove the `authRefresh` call, the data I'm retrieving (`"foo"`) is getting back every single time it seems, even at this refresh pace)
I also tried both with `npm run dev` and `npm run build` + `npm run preview`. In the second case I'm logged out faster. That's because `npm run dev` is slower of course.

**Reply by ganigeorgiev** [reply]
Could you share your code sample where you are authenticating? I just don't know how to reproduce it.

**Reply by peterhijma** [reply]
```js
async function login({ email, password }) {
    await $pb.collection("users").authWithPassword(email, password).then(async (user) => {
      await navigateTo("/foo");
    })
}
```

It's in a separate page and is a very simple form with this as submit handler.

**Reply by peterhijma** [reply]
@ganigeorgiev If you'd like I can create a full example. What's the best way to do this for you?

I could do it tomorrow.

**Reply by ganigeorgiev** [reply]
I don't think this is related to PocketBase and most likely either the plugins approach is not the correct one for Nuxt3 OR `useCookie` fails to return a cookie.

I'll try to investigate it sometime later but I'll have to go through the Nuxt3 documentation to research in more details when and where the plugins run.

**Comment by ganigeorgiev** [comment]
@peterhijma I was able to reproduce it bit I'm not sure why it is happening.

Also note that the issue happens only when the client-side part runs. For example if you refresh with `view-source:http://localhost:3000` to avoid running the client-side js, on the server side it works fine.

Additionally, if you `console.log(err.originalError)` in the `authRefresh` try/catch, you'll get something like:

```
TypeError: Failed to fetch
    at e.<anonymous> (:3000/_nuxt/entry.17f65496.js:3:45102)
    at :3000/_nuxt/entry.17f65496.js:3:3511
    at Object.next (:3000/_nuxt/entry.17f65496.js:3:3616)
    at :3000/_nuxt/entry.17f65496.js:3:2532
    at new Promise (<anonymous>)
    at ce (:3000/_nuxt/entry.17f65496.js:3:2269)
    at e.send (:3000/_nuxt/entry.17f65496.js:3:43444)
    at t.authRefresh (:3000/_nuxt/entry.17f65496.js:3:26381)
    at :3000/_nuxt/entry.17f65496.js:3:47176
    at Si (:3000/_nuxt/entry.17f65496.js:1:72768)
```

Which is certainly strange.

Unfortunately currently I don't have the time to go down the rabbit hole to fully understand what is causing it and whether it is nuxt, browser or something with the JS SDK (_although I doubt the last one because we use whatever global `fetch` implementation is available for sending the request_).

One hacky workaround for the time being could be to just ignore the `0` status code errors only on the client-side. You can try to change the `authRefresh()` try/catch block with:

```js
try {
    // get an up-to-date auth store state by verifying and refreshing the loaded auth model (if any)
    pb.authStore.isValid && await pb.collection('users').authRefresh();
} catch (err) {
  // ignore the error if it is in the browser AND it has status code 0 (usually indicating non-server error, like autoCancellation, browser request abort due to navigation, etc.)
  if (!process.client || err.status != 0) {
    // clear the auth store on failed refresh
    pb.authStore.clear();
  }
}
```

**Reply by peterhijma** [reply]
Thanks for trying to reproduce and succeeding 🥳 The workaround is good enough. I'll also try to do some more deep diving I find out why it happens. I'll let you know if I find anything.
